LibraBFT/Impl/Consensus/EpochManager.agda

```diff
@@ -0,0 +1,499 @@
+
+{- Byzantine Fault Tolerant Consensus Verification in Agda, version 0.9.
+   Copyright (c) 2021, Oracle and/or its affiliates.
+   Licensed under the Universal Permissive License v 1.0 as shown at https://opensource.oracle.com/licenses/upl
+-}
+
+open import LibraBFT.Base.PKCS
+open import LibraBFT.Base.Types
+import      LibraBFT.Impl.Consensus.BlockStorage.BlockStore          as BlockStore
+open import LibraBFT.Impl.Consensus.EpochManagerTypes
+import      LibraBFT.Impl.Consensus.Liveness.ExponentialTimeInterval as ExponentialTimeInterval
+import      LibraBFT.Impl.Consensus.Liveness.RoundState              as RoundState
+import      LibraBFT.Impl.Consensus.MetricsSafetyRules               as MetricsSafetyRules
+import      LibraBFT.Impl.Consensus.RoundManager                     as RoundManager
+import      LibraBFT.Impl.Consensus.SafetyRules.SafetyRulesManager   as SafetyRulesManager
+import      LibraBFT.Impl.Consensus.TestUtils.MockStorage            as MockStorage
+import      LibraBFT.Impl.OBM.ECP-LBFT-OBM-Diff.ECP-LBFT-OBM-Diff-1  as ECP-LBFT-OBM-Diff-1
+open import LibraBFT.Impl.OBM.Logging.Logging
+import      LibraBFT.Impl.OBM.Rust.Duration                          as Duration
+open import LibraBFT.Impl.OBM.Rust.RustTypes
+import      LibraBFT.Impl.Storage.DiemDB.DiemDB                      as DiemDB
+import      LibraBFT.Impl.Types.BlockInfo                            as BlockInfo
+import      LibraBFT.Impl.Types.EpochChangeProof                     as EpochChangeProof
+import      LibraBFT.Impl.Types.ValidatorVerifier                    as ValidatorVerifier
+open import LibraBFT.Impl.Types.Verifier
+open import LibraBFT.ImplShared.Consensus.Types
+open import LibraBFT.ImplShared.Interface.Output
+open import LibraBFT.ImplShared.Util.Util
+open import LibraBFT.Prelude
+open import Optics.All
+------------------------------------------------------------------------------
+import      Data.String                                              as String
+
+module LibraBFT.Impl.Consensus.EpochManager where
+
+open import LibraBFT.ImplShared.Util.Dijkstra.RWST        public
+open import LibraBFT.ImplShared.Util.Dijkstra.RWST.Syntax public
+
+------------------------------------------------------------------------------
+
+data RlecState : Set where
+  RSNothing  : RlecState
+  RSNeedECP  : ReconfigEventEpochChange → RlecState
+  RSNeedRLEC : EpochChangeProof         → RlecState
+
+data ProcessMessageAction : Set where
+  PMContinue        : ProcessMessageAction
+  PMInput           : Input            → ProcessMessageAction
+  PMNewEpochManager : EpochManager     → List Output    → ProcessMessageAction
+  PMSendECP         : EpochChangeProof → AccountAddress → Author {-Text-} → Epoch → Round → ProcessMessageAction
+  PMSendEpochRRq    : EpRRqWire        → AccountAddress → ProcessMessageAction
+
+data PMOutput : Set where
+  PMErr  : ErrLog  → PMOutput
+  PMInfo : InfoLog → PMOutput
+
+EM : Set → Set₁
+EM = RWST Unit PMOutput RlecState
```

Comment:
`RWST` is parameterized:
- `data RWST (Ev Wr St : Set) : Set → Set₁ where`

`LBFT` gives the following parameters:
- `LBFT = RWST Unit Output RoundManager`

`EM` gives the following parameters:
- `EM = RWST Unit PMOutput RlecState`

Note that whatever `Ev` is instantiated is then used in a `List` because:
- `  RWST-tell : List Wr → RWST Ev Wr St Unit`

Note: that in a "real" `RWST`, the type of `Ev` is not a list,
instead it must be a monoid, so when `tell` is used it uses the monoid `<>` operator
to combine the existing stuff in the writer with the new addition.

Does that satisfy your curiosity?

Reply:
Yes, except that I think you misspelled `Wr` (twice) :smile:.  Thanks.